File: Chapter 1/Section 1.3/namenum/namenum.py
"""
ID: deadly51
TASK: namenum
LANG: PYTHON3
"""
with open("namenum.in", "r") as file:
    number = file.readline().strip()
    size = len(number)
with open("dict.txt") as file:
    dicti = [x.strip() for x in file.readlines()]

letters = {2: "ABC", 5: "JKL", 8: "TUV",
          3: "DEF", 6: "MNO", 9: "WXY",
          4: "GHI", 7: "PRS"}


words = []
def func(name, number, i):
    if i == len(number):
        for j in dicti:
            if name == j:
                words.append(name)
        return
    
    for z in letters[int(number[i])]:
        temp = name+z
        for j in dicti:
            if j.startswith(temp):
                func(temp, number, i+1)
                break


func("", number, 0)
if not words:
    words.append("NONE")
with open("namenum.out", "w") as file:
    for i in range(len(words)):
        file.write(f"{words[i]}\n")

File: Chapter 1/Section 1.3/namenum/namenum.cpp

/*
ID: sagosag1
TASK: namenum
LANG: C++
*/

#include <bits/stdc++.h>

using namespace std;

int main() {
	ifstream fin("namenum.in");
	ofstream fout("namenum.out");
	ifstream dict("dict.txt");

	string num; fin >> num;
	bool found = false;

	string name;
	while (dict >> name) {
		if (name.length() == num.length()) {
			string res = "";
			for (auto& c: name) {
				if (c == 'Q' || c == 'Z') {
					break;
				}
				if (c >= 'A' && c < 'Q') {
					res += ((c - 'A') / 3) + '2';
				} else if (c > 'Q' && c < 'Z'){
					res += ((c - 'Q') / 3) + '7';
				} else {
					break;
				}
			}
			if (res != num) {
				continue;
			}
			found = true;
			fout << name << '\n';
		}
	}

	if (!found) {
		fout << "NONE\n";
	}

	return 0;
}

